### PR TITLE
Add a guide howto deploy a Quarkus app to Heroku

### PR DESCRIPTION
**Pull request by michael-simons**
This is a fully tested guide how to deploy a Quarkus app to Heroku.com in response to @maxandersen 

https://twitter.com/maxandersen/status/1393575049295245314?s=20

Feel free to fix my worst typos please (Not an English native speaker).

### CODE REVIEW

File: docs/src/main/asciidoc/deploying-to-heroku.adoc [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+    -DprojectGroupId=org.acme \
+    -DprojectArtifactId=code-with-quarkus-on-heroku \
+    -DclassName="org.acme.getting.started.GreetingResource" \
+    -Dpath="/hello"
+cd code-with-quarkus-on-heroku
+git init && git add . && git commit -am "Create initial project."
+----
+
+Heroku can react on changes in your repository, run CI and redeploy your application when your code changes.
+Therefor we start with a valid repository already.
```
Comment (gsmet) [inline docs/src/main/asciidoc/deploying-to-heroku.adoc:67]:
```suggestion
Therefore we start with a valid repository already.
```

File: docs/src/main/asciidoc/deploying-to-heroku.adoc [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+
+Also, make sure you're Heroku CLI is working:
```
Comment (gsmet) [inline docs/src/main/asciidoc/deploying-to-heroku.adoc:69]:
```suggestion
Also, make sure your Heroku CLI is working:
```

File: docs/src/main/asciidoc/deploying-to-heroku.adoc [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+
+[source,baseh]
```
Comment (gsmet) [inline docs/src/main/asciidoc/deploying-to-heroku.adoc:71]:
```suggestion
[source,bash]
```

File: docs/src/main/asciidoc/deploying-to-heroku.adoc [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+----
+heroku --version
+heroku login
+----
+
+== Prepare the Quarkus HTTP Port
+
+Heroku picks a random port and assigns it to the container that is eventually running your Quarkus application.
+That port is available as environment variable under `$PORT`.
```
Comment (gsmet) [inline docs/src/main/asciidoc/deploying-to-heroku.adoc:80]:
```suggestion
That port is available as an environment variable under `$PORT`.
```

File: docs/src/main/asciidoc/deploying-to-heroku.adoc [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+[source,bash]
+----
+echo "quarkus.http.port=\${PORT:8080}" >> src/main/resources/application.properties
+git commit -am "Configure the HTTP Port."
+----
+
+== Deploy the repository and build on Heroku
+
+The first variant uses the Quarkus Maven build to create the _quarkus-app_ application structure containing the runnable "fast-jar" as well as all libraries needed
+inside Heroku's build infrastructure and than deploying that result, the other one uses a local build process to create optimized container.
```
Comment (gsmet) [inline docs/src/main/asciidoc/deploying-to-heroku.adoc:100]:
```suggestion
inside Heroku's build infrastructure and then deploying that result, the other one uses a local build process to create an optimized container.
```

File: docs/src/main/asciidoc/deploying-to-heroku.adoc [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+
+Two additional files are needed in your applications root directory:
```
Comment (gsmet) [inline docs/src/main/asciidoc/deploying-to-heroku.adoc:102]:
```suggestion
Two additional files are needed in your application's root directory:
```

File: docs/src/main/asciidoc/deploying-to-heroku.adoc [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+
+To access the REST endpoint via curl, run:
+
+[source,bash]
+----
+APP_NAME=`heroku info | grep  "=== .*" |sed "s/=== //"`
+curl $APP_NAME.herokuapp.com/hello
+----
+
+Of course, you can use the Heroku CLI to connect this repo to your GitHub account, too, but this is out of scope for the guide.
```
Comment (gsmet) [inline docs/src/main/asciidoc/deploying-to-heroku.adoc:146]:
```suggestion
Of course, you can use the Heroku CLI to connect this repo to your GitHub account, too, but this is out of scope for this guide.
```

File: docs/src/main/asciidoc/deploying-to-heroku.adoc [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+
+== Deploy as container
+
+The advantage of pushing a whole container is that we are in complete control over its content and maybe even chose to deploy a container with a native binary running on GraalVM.
```
Comment (gsmet) [inline docs/src/main/asciidoc/deploying-to-heroku.adoc:150]:
```suggestion
The advantage of pushing a whole container is that we are in complete control over its content and maybe even choose to deploy a container with a native executable running on GraalVM.
```

File: docs/src/main/asciidoc/deploying-to-heroku.adoc [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+We need to add an extension to our project to build container images via the Quarkus Maven plugin:
+
+[source,bash]
+----
+mvn quarkus:add-extension -Dextensions="container-image-docker"
+git add pom.xml
+git commit -am "Add container-image-docker extension."
+----
+
+The image we are gonna build needs to named accordingly to work with Herokus registry and deployment.
```
Comment (gsmet) [inline docs/src/main/asciidoc/deploying-to-heroku.adoc:168]:
```suggestion
The image we are going to build needs to be named accordingly to work with Heroku's registry and deployment. 
```

File: docs/src/main/asciidoc/deploying-to-heroku.adoc [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+
+With Docker installed, you can now push the image and release it:
+
+[source,bash]
+----
+docker push registry.heroku.com/$APP_NAME/web
+heroku container:release web --app $APP_NAME
+----
+
+You can and should check the logs to see if your application is now indeed running from the container
```
Comment (gsmet) [inline docs/src/main/asciidoc/deploying-to-heroku.adoc:189]:
```suggestion
You can and should check the logs to see if your application is now indeed running from the container:
```

File: docs/src/main/asciidoc/deploying-to-heroku.adoc [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+
+[source,bash]
+----
+heroku logs --app $APP_NAME --tail
+----
+
+The initial push is rather big, as all layers of the image need to be transferred.
+All following will be smaller.
```
Comment (gsmet) [inline docs/src/main/asciidoc/deploying-to-heroku.adoc:197]:
```suggestion
The following pushes will be smaller.
```

File: docs/src/main/asciidoc/deploying-to-heroku.adoc [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+
+The biggest advantage we take when deploying our app as a container is to deploy a container with the natively compiled application.
+Why? Because Heroku will stop or sleep the application when there's no incoming traffic.
+A native application will wake much faster from its sleep.
```
Comment (gsmet) [inline docs/src/main/asciidoc/deploying-to-heroku.adoc:201]:
```suggestion
A native application will wake up much faster from its sleep.
```

**Review by gsmet — CHANGES_REQUESTED** [review]
Nice. I spotted a couple of typos and suggested a few improvements. Can you have another look?

**Comment by gsmet** [comment]
Just saw, we had your blessings to fix the typos so I squashed the commits and rebased. Let's wait for CI.

**Review by gsmet — APPROVED** [review]

**Comment by gsmet** [comment]
Thanks!

**Comment by michael-simons** [comment]
Very much welcome! It was an overall pleasure deploying my first Quarkus GraphQL app to Heroku, so thank you!

**Pull request merged**